ludwig/modules/lr_scheduler.py

```diff
@@ -0,0 +1,161 @@
+
+from ludwig.constants import MINIMIZE, TRAINING, VALIDATION
+from ludwig.modules.metric_modules import get_metric_cls
+from ludwig.schema.lr_scheduler import LRSchedulerConfig
+from ludwig.utils.metric_utils import TrainerMetric
+from ludwig.utils.trainer_utils import ProgressTracker
+
+
+class ReduceLROnPlateauLimited(ReduceLROnPlateau):
+    def __init__(self, optimizer: Optimizer, mode: str, reduce_limit: int, factor: float, patience: int):
+        super().__init__(optimizer, mode=mode, factor=factor, patience=patience)
+        self.reduce_limit = reduce_limit
+        self._num_reduce_lr = 0
+
+    def step(self, metrics, epoch=None):
+        if self._num_reduce_lr >= self.reduce_limit:
+            # Already reduce the LR as many times as we will allow
+            return
+
+        return super().step(metrics, epoch=epoch)
+
+    def _reduce_lr(self, epoch):
+        super()._reduce_lr(epoch)
+        self._num_reduce_lr += 1
+
+
+class LRScheduler:
+    def __init__(
+        self,
+        config: LRSchedulerConfig,
+        optimizer: Optimizer,
+        steps_per_checkpoint: int = 1000,
+        total_steps: int = 10000,
+    ):
+        self.config = config
+        self.optimizer = optimizer
+        self.validation_metric = get_metric_cls(self.config.reduce_eval_metric)
+
+        # Scheduler updated each training step
+        self.step_info = StepInfo(steps_per_checkpoint, total_steps, self.config)
+        self._train_scheduler = get_schedule_with_warmup(self.config, self.optimizer, self.step_info)
+
+        # Scheduler updated each eval step
+        self._eval_scheduler = None
+        if self.config.reduce_on_plateau > 0:
+            mode = "min" if self.validation_metric.get_objective() == MINIMIZE else "max"
+            self._eval_scheduler = ReduceLROnPlateauLimited(
+                optimizer=self.optimizer,
+                mode=mode,
+                reduce_limit=self.config.reduce_on_plateau,
+                factor=self.config.reduce_on_plateau_rate,
+                patience=self.config.reduce_on_plateau_patience,
+            )
+
+        self.reset(steps_per_checkpoint, total_steps)
+
+    def reset(self, steps_per_checkpoint: int, total_steps: int):
+        # Retain state but update number of steps for training
+        self.step_info.reset(steps_per_checkpoint, total_steps)
+
+    def step(self):
+        self._train_scheduler.step()
+
+    def eval_step(self, progress_tracker: ProgressTracker, validation_field: str):
+        if self._eval_scheduler is None:
+            # No reduce on plateau
+            return
+
+        if self.config.reduce_eval_split == TRAINING:
+            split_metrics = progress_tracker.train_metrics
+        elif self.config.reduce_eval_split == VALIDATION:
+            split_metrics = progress_tracker.validation_metrics
+        else:  # if self.config.reduce_eval_split == TEST:
+            split_metrics = progress_tracker.test_metrics
```

Comment:
Not in scope for this PR, but we could/should be more prescriptive about which split to use for reduction (training or validation). Using `TEST` doesn't seem quite right. CC: @dantreiman @w4nderlust

Reply:
Agreed test seems like a violation of the concept of a test set. I think training or validation both have merits, but usually you'd want to use training, I'd say.